### PR TITLE
[FLINK-28781][network] Hybrid Shuffle should support compression

### PR DESCRIPTION
**Pull request by reswqa**
## What is the purpose of the change

*Compression is a useful feature for batch jobs, which can significantly reduce disk load and the amount of data transferred over the network. Hybrid shuffle should also support the compression of spilled data, especially under the full spilling strategy.*

## Brief change log

  - *Introduce compression support for HsMemoryDataSpiller*

## Verifying this change

This change added tests in HsMemoryDataSpillerTest and HsSubpartitionFileReaderImplTest.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/configuration/NettyShuffleEnvironmentOptions.java [hunk 1]
```diff
@@ -70,20 +70,20 @@ public class NettyShuffleEnvironmentOptions {
...
      * more effective for IO bounded scenario when data compression ratio is high.
      */
     @Documentation.Section(Documentation.Sections.ALL_TASK_MANAGER_NETWORK)
-    public static final ConfigOption<Boolean> BLOCKING_SHUFFLE_COMPRESSION_ENABLED =
-            key("taskmanager.network.blocking-shuffle.compression.enabled")
+    public static final ConfigOption<Boolean> BATCH_SHUFFLE_COMPRESSION_ENABLED =
+            key("taskmanager.network.batch-shuffle.compression.enabled")
                     .booleanType()
                     .defaultValue(true)
+                    .withFallbackKeys("taskmanager.network.blocking-shuffle.compression.enabled")
```
Comment (xintongsong) [inline flink-core/src/main/java/org/apache/flink/configuration/NettyShuffleEnvironmentOptions.java:83]:
```suggestion
                    .withDeprecatedKeys("taskmanager.network.blocking-shuffle.compression.enabled")
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/HsMemoryDataSpiller.java [hunk 2]
```diff
@@ -60,13 +64,17 @@ public class HsMemoryDataSpiller implements AutoCloseable {
     /** File channel to write data. */
     private final FileChannel dataFileChannel;
 
+    private final BufferCompressor bufferCompressor;
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/HsMemoryDataSpiller.java:67]: Should mark `@Nullable`. Same for the corresponding constructor parameter.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionType.java [hunk 3]
```diff
@@ -196,4 +196,8 @@ public boolean isBounded() {
     public boolean isPersistent() {
         return isPersistent;
     }
+
+    public boolean canBeCompressed() {
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionType.java:200]:
```suggestion
    public boolean supportCompression() {
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/HsMemoryDataSpiller.java (hunk 2 above)
Comment (reswqa) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/HsMemoryDataSpiller.java:67]: marked.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionType.java (hunk 3 above)
Comment (reswqa) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionType.java:200]: Renamed.

**Comment by reswqa** [comment]
@xintongsong Thanks for the review, I have address all of comments, please take look.

**Review by xintongsong — APPROVED** [review]
LGTM. Merging.

**Pull request closed without merging**